Migrate testing to pytest

emirvine opened the pull request:
**Description:** 
Porting testing to native pytest.
These changes were done on a Windows machine in python 3.5.
Results in 55 passed, 4 skipped, 4 xfailed tests. Total 80% coverage.

**Motivation and context:**
Some tests were running into the depreciation warning when running (nose) tests with pytest.
Here, we make use of pytest fixtures and parametrize and monkeypatch to simplify the tests
and lose the mock dependency.
Addresses #97

**Types of changes:**
Tests are all now in pytest format.
Some small formatting changes were included to make these tests and some functions they 
use to be in line with PEP8.

yarikoptic commented:
Thank you @emirvine very much for your help.
Do you want us to take over from here or you would like me to review the changes/do recommendations so you could finalize it yourself?  either way would work for us.

Cheers

emirvine commented:
I would like to go through the review process if that's alright with you. But I do have my specialist coming up and so will be busier with that for the next two weeks. If this is something that you want to finish quickly then it might be better to take it from here without me.

yarikoptic commented:
ok -- let me then fixup for some "corner cases" (like that stub.py file which should just get copied/removed within some global fixture) to get ball rolling and then review and see how it goes.  There is no urge ATM since not that much of development is happening ATM, so less of a chance of conflict.

emirvine commented:
Great, sounds good.

yarikoptic commented:
woohoo -- I have fixed up few loose ends (also seems have unraveled some prior non-effective test, fixed for that in the last commit) and hopefully now would get time to review the rest

yarikoptic commented on duecredit/tests/test_api.py:
```diff
@@ -41,102 +87,107 @@ def purpose_of_life():
-
-
-def check_noincorrect_import_if_no_lxml_numpy(kwargs, env):
+        pytest.xfail("Fails for some reason on Windows")
```
have no windows env handy -- could you please check if it is still the case -- may be we fixed for it already

yarikoptic commented on duecredit/tests/test_api.py:
```diff
@@ -41,102 +87,107 @@ def purpose_of_life():
+        pytest.skip("We need to have numpy to test correct operation")
 
+    if on_windows:
+        pytest.xfail("Fails for some reason on Windows")
```
here as well -- would be nice if windows situation has improved ;)

yarikoptic commented on duecredit/tests/test_collector.py:
```diff
@@ -112,126 +107,126 @@ def test_dcite_method():
-
+            assert citation.cite_module
 
         class SomeClass2(object):
```
please add a comment that this class is not directly used and just there to test for that ;)

yarikoptic commented on duecredit/tests/test_collector.py:
```diff
@@ -112,126 +107,126 @@ def test_dcite_method():
             def method2(self, arg1, kwarg2="blah"):
-                assert_equal(arg1, "magical")
+                assert arg1 == "magical"
                 return "load"
```
since not used really please replace the body with
```python
pass  # pragma: no cover
```
to signal coverage that we do not want this covered ;-)

yarikoptic commented on duecredit/tests/test_dueswitch.py:
```diff
@@ -7,26 +7,35 @@
-    # injector.activate()
+
+def test_dueswitch_activate(monkeypatch):
+    if due.active:
```
please add `# pragma: no cover` here as well

emirvine commented:
I pushed a new commit that addresses most of your comments. 

In Windows, os.environ doesn't have PYTHONPATH. It's possible to do this by monkeypatching to sys.path, but since the test uses subprocesses this wouldn't work. I'm not sure how Windows deals with PYTHONPATH.

I'm not sure how `# pragma: no cover` works, and which specific lines you are referring to. Feel free to just add them in where you see fit or let me know what line number and file to add them.

yarikoptic commented:
`  # pragma: no cover` says that this block does not have to be tested.  If it is added on a simple statement line -- just that line coverage is not demanded.  if added at a line defining a new block (such as `if`, `for`, `def` etc) -- the entire block does not have to be covered. e.g. we have already
```
$> git grep -A2 'no cover' 
duecredit/injections/injector.py:        if not self._active:  # pragma: no cover
duecredit/injections/injector.py-            lgr.error("Must have not happened, but we will survive!")
duecredit/injections/injector.py-        lgr.debug("Assigning original importer")
--
duecredit/utils.py:except:  # pragma: no cover
duecredit/utils.py-    on_debian_wheezy = False
duecredit/utils.py-
```

so in the first case -- the `if` statement including `lgr.error` invocation does not have to be tested.  In the second -- exception handling block.

> In Windows, os.environ doesn't have PYTHONPATH

it might not be there by default, but if defined, I think it could still be 'used'. see e.g. http://stackoverflow.com/questions/3701646/how-to-add-to-the-pythonpath-in-windows-7

To ease us moving forward I will merge this PR now since we are on the right track ;-).  Outstanding TODOs would be for us to
- [ ] investigate why some coverage went down: use this report for reference: https://codecov.io/gh/duecredit/duecredit/compare/2221bfdb3c89afd30daac78de1a78a893e932b58...2db18b89c6fa3bc8cba3c836b2b2a21166cb989e/changes    we need to at least return back covering that `Citation.__repr__`
- [ ] add a few `no cover` statements
